mayastor/src/bdev/dev/nvmx/controller.rs

```diff
@@ -236,155 +242,224 @@ impl<'a> NvmeController<'a> {
                     source: Errno::EBUSY,
                 });
             }
-            _ => {}
         }
 
+        info!(
+            "{} initiating controller reset, failover = {}",
+            self.name, failover
+        );
+
         if failover {
             warn!(
                 "{} failover is not supported for controller reset",
                 self.name
             );
         }
 
-        let reset_ctx = RESET_CTX_POOL
-            .get()
-            .unwrap()
-            .get(ResetCtx {
-                name: self.name.clone(),
-                cb,
-                cb_arg,
-                spdk_handle: self.ctrlr_as_ptr(),
-            })
-            .ok_or(CoreError::ResetDispatch {
-                source: Errno::ENOMEM,
-            })?;
+        let io_device = Arc::clone(&self.inner.as_ref().unwrap().io_device);
+        let reset_ctx = ResetCtx {
+            name: self.name.clone(),
+            cb,
+            cb_arg,
+            spdk_handle: self.ctrlr_as_ptr(),
+            io_device,
+        };
 
         // Mark controller as being under reset and schedule asynchronous reset.
-        self.set_state(NvmeControllerState::Resetting);
+        self.reset_active = true;
```

Comment:
ah! any reason for this?

Reply:
Yes.
I decided to get rid of Resetting as a separate state to simplify checking against controller availability in other code that makes judgements about availability of the controller.
Its much easier to check against ::Running as a valid state then ::Running && ::Resetting, because ::Resetting means 'Controller is OK and just being reset'.
So reset feels to me like an action inside ::Running state.

Reply:
Yes, I understand your thinking, and it makes sense. But at the same time, it also seems to lead to situations that function like `.state()` or `.set_state()` would not really tell you the whole picture when using these functions and their respective return types :thinking: 

So when we are resetting, we have the controller's global state (which is guarded by the mutex) and we keep `self.rest_active == true` until we have iterated all cores with an io channel to the device, right? 

If so we can perhaps introduce an extra variant of the enum type for example:

```rust
//to make it more clear for ourselves
type resetting: bool; 

pub enum NvmeControllerState {
 /// the old states we had and change Running too:
 Running(resetting)
}
```

This would  allow when calling `state()`  for the callee to really understand whats going on i.e:
```rust
match self.state() {
   Running(restting) if !resetting => {  // we are running and not resetting }, 
   Running(restting) if resetting => {  // guess what...   },
}
```

For cases where we don't care we can do:

```rust 
 match self.state() {
    Running(_) => {  // we are running but we don't care what state  we are in }, 
}
``` 

As this is somewhat verbose we can make use of the `matches!()` macro that is available in our version of rust: 
```rust
if matches!(self.state(), Running(_)) { } 
```

So this sounds like an awful lot like bikeshedding, but there is reason to the madness. We have a pretty messy state assessment within the current code base, and I would like to avoid that if at all possible. For example, we set a state from A to B but forgot to set a bool to false, making the code go poof.  

So what do you think? Is it worth trying it out such that we can differentiate internally and have just a single variable (and method) to consult to get the fully defined state rather than state + some other variable to consult? Then again, I might be pedantic :v:

Reply:
Interesting idea, I like it.
So we basically can hide sub-states for Running state inside the state itself.
We should definitely give it a try, so I will update the PR to reflect that.
Thanks !